[DOCS] Separate "user lookup" into its own doc (#88533)
elastic/elasticsearch: x-pack/docs/en/security/authentication/user-lookup.asciidoc

[role="xpack"]
[[user-lookup]]
=== Looking up users without authentication

{es} <<realms,realms>> exist primarily to support
<<setting-up-authentication,user authentication>>. 
Some realms authenticate users with a password (such as the
<<native-realm,`native`>> and <<ldap-realm,`ldap`>> realms), and other realms use
more complex authentication protocols (such as the <<saml-realm,`saml`>> and
<<oidc-realm,`oidc`>> realms).
In each case, the _primary_ purpose of the realm is to establish the identity of
the user who has made a request to the {es} API.

However, some {es} features need to _look up_ a user without using their credentials.

- The <<run-as-privilege,`run_as`>> feature executes requests on behalf of
  another user. An authenticated user with `run_as` privileges can perform
  requests on behalf of another unauthenticated user.

- The <<authorization_realms,delegated authorization>> feature links two realms
  together so that a user who authenticates against one realm can have the roles
  and metadata associated with a user from a different realm.

In each of the above cases the second user must be resolved without directly

> authenticating. The ability for a realm to lookup a user by username is an

Reviewer: ```suggestion
authenticating. The ability for a realm to look up a user by username is an
```

[role="xpack"]
[[user-lookup]]
=== Looking up users without authentication

{es} <<realms,realms>> exist primarily to support
<<setting-up-authentication,user authentication>>. 

> Some realms authenticate users via a password (for example the
> <<native-realm,`native`>> and <<ldap-realm,`ldap`>> realms) and other realms use
> more complex authentication protocols (for example the <<saml-realm,`saml`>> and
> <<oidc-realm,`oidc`>> realms).

Reviewer: ```suggestion
Some realms authenticate users with a password (such as the
<<native-realm,`native`>> and <<ldap-realm,`ldap`>> realms), and other realms use
more complex authentication protocols (such as the <<saml-realm,`saml`>> and
<<oidc-realm,`oidc`>> realms).
```

In each case, the _primary_ purpose of the realm is to establish the identity of
the user who has made a request to the {es} API.

However, some {es} features need to _lookup_ a user without using their credentials.

- The <<run-as-privilege,`run_as`>> feature executes requests on behalf of
  another user. An authenticated user with `run_as` privileges can perform
  requests on behalf of another unauthenticated user.

- The <<authorization_realms,delegated authorization>> feature links two realms
  together so that a user who authenticates against one realm can have the roles
  and metadata associated with a user from a different realm.

> In each of the above cases the second user must be resolved without directly

Reviewer 2: Suggestion for a bit more clarity..

In each of the above cases a user must first authenticate to one realm, and then authorization is handled by looking up that user in a second realm. The user is authenticated in the first realm, but only must exist in the second realm (i.e. the user does not credentials to the second realm). Querying the second realm for authorization purposes is known as user lookup. 

See the <<run-as-privilege,`run_as`>> and <<authorization_realms,delegated authorization>> documentation for which realms support the ability to perform run_as or delegate authorization. In both cases the only following realms can be for the user lookup.

Reviewer 2: resolving as superseded by Adam's suggestion below.

authenticating. The ability for a realm to lookup a user by username is an

> optional operation that is not supported by all realms.

Reviewer: I like Jake's suggestion, so I added it as a suggested change with minor edits.

> optional operation that is not supported by all realms.

Reviewer 2: Thanks Adam ... i really need to figure out how to do multi-line suggestions in GH (and edits look good)

> optional operation that is not supported by all realms.

Author: I think the phrasing around "Authorization" is slightly misleading for run-as.

Authorization is performed for the first user to ensure they are permitted to "run as" the second user, and then authorization checks are performed again to ensure the 2nd user is permitted to execute the request.  
The proposed wording implies that authorization is exclusive performed against the second user.

I'll try and come up with a variation that avoid that implication.

> optional operation that is not supported by all realms.

Reviewer: @tvernum, here's a suggested version that incorporates the distinction you highlighted:

> In both cases, a user must authenticate to a realm first. After successfully authenticating, two authorization 
>checks occur:
> * The first check ensures that the authenticated user has privileges to execute requests as a different user.
> * The second check ensures that the unauthenticated user (the second user) has privileges to execute the 
>specified request.
> 
> Querying the second realm to retrieve the unauthenticated user's privileges is known as a _user lookup_. The authenticated user doesn't need credentials in the second realm; they only need to exist in that realm.

> optional operation that is not supported by all realms.

Reviewer 3: I think describing it in order of operations helps to understand the relationship of delegated authz vs run-as.

If a realm successfully authenticates a user, a realm can be configured to choose the authenticated user's privileges by one of these two configured options. The options cannot be used simultaneously.
1. The realm that performed authentication can do the authorization lookups.
2. The realm that performed authentication can list a subset of other realms to do the authorization lookups by username.

After the authenticated user has been authorized by one of the above configured methods, one more check is done before servicing a request.

If the request includes a run-as header, the authenticated user's privileges are checked to see if it is allowed to run-as that requests user. If yes, all configured realms are asked to do a lookup of the run-as user by username. If the run-as user it found, the lookup realm returns the run-as user with a set of privileges versus the original authenticated user. The request is then executed as the run-as user, with the run-as user's permissions.

All 10 realms can request delegated authorization: reserved, native, file, ldap, ad, pki, saml, oidc, kerberos, jwt.

Only 4 realms can fulfill delegated authorizations: native, file, ldap, ad 

There is overlap where one LDAP/AD realm can authenticate, and then request delegated authorization from a different LDAP/AD realm. It does not make sense for an LDAP/AD realm to delegate authorization to itself, just omit the delegated authorization setting.

> optional operation that is not supported by all realms.

Author: This is a doc trying to describe user lookup, not run as or delegated authorization. Both those features have their own docs, I don't want to explain them again here.

The point of mentioning them is to provide examples of why the concept of user lookup exists so that the description has some context, but if the readers want to know more details about run-as or delegated authorization they should read the pages for those 2 features.

I'm happy to make this paragraph clearer, but I don't want to end up becoming a distraction from the purpose of the page.

> The reserved, <<native-realm,`native`>> and <<file-realm,`file`>> realms always
> support user lookup.
> 
> The <<ldap-realm,`ldap`>> realm supports user lookup when the realm is configured
> in <<ldap-realm-configuration,_user search_ mode>>. User lookup is not support
> when the realm is configured with `user_dn_templates`.
> 
> User lookup support in the <<active-directory-realm,`active_directory`>> realm
> requires that the realm be configured with a <<ref-ad-settings,`bind_dn`>> and a
> bind password.

Reviewer: ```suggestion
* The reserved, <<native-realm,`native`>> and <<file-realm,`file`>> realms always 
support user lookup.
* The <<ldap-realm,`ldap`>> realm supports user lookup when the realm is configured
in <<ldap-realm-configuration,_user search_ mode>>. User lookup is not support
when the realm is configured with `user_dn_templates`.
* User lookup support in the <<active-directory-realm,`active_directory`>> realm
requires that the realm be configured with a <<ref-ad-settings,`bind_dn`>> and a
bind password.
```

Reviewer: I think that using a bulleted list to group these items would be good for readability.

The `pki`, `saml`, `oidc`, `kerberos` and `jwt` realms do not support user
lookup.

> NOTE: If you want to use a realm for _user lookup_ only, and prevent users from 
>       authenticating against that realm, you can
>       <<ref-realm-settings,configure the realm>> and set
>       `authentication.enabled` to `false`

Reviewer: ```suggestion
NOTE: If you want to use a realm only for user lookup and prevent users from 
authenticating against that realm, you can <<ref-realm-settings,configure the realm>>
and set `authentication.enabled` to `false`
```

The user lookup feature is an internal capability that is used to implement the
`run-as` and delegated authorization features - there are no APIs for user lookup.
If you wish to test your user lookup configuration, then you can do this with
`run_as`. Use the <<security-api-authenticate>> API, authenticate as a

> `superuser` (e.g. the builtin `elastic` user) and specify the

Author: I don't think so. It's the only user we ship with that is able to be used for this sort of testing, so I don't think there's any reason so avoid suggesting it as an option here.
